### PR TITLE
[RFC]: Add support for drain operation on compress streams

### PR DESCRIPTION
**Pull request by LaurentiuM1234**
This is a rough attempt of adding support for the drain operation on compress streams.

### CODE REVIEW

File: sound/soc/sof/compress.c [hunk 1]
```diff
@@ -287,6 +287,9 @@ static int sof_compr_trigger(struct snd_soc_component *component,
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
 		stream.hdr.cmd |= SOF_IPC_STREAM_TRIG_RELEASE;
 		break;
+	case SNDRV_PCM_TRIGGER_DRAIN:
+		stream.hdr.cmd |= SOF_IPC_STREAM_TRIG_DRAIN;
+		break;
```
Comment (plbossart) [inline sound/soc/sof/compress.c:292]: you probably need a different behavior for capture and playback.

From the alsa-lib PCM documentation https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#ga49afc5b8527f30c33fafa476533c9f86

"
For playback wait for all pending frames to be played and then stop the PCM. For capture stop PCM permitting to retrieve residual frames.
"

File: sound/soc/sof/ipc3.c [hunk 2]
```diff
@@ -898,6 +898,35 @@ static void ipc3_xrun(struct snd_sof_dev *sdev, u32 msg_id)
...
+	snd_sof_ipc_msg_data(sdev, stream->substream, &drain, sizeof(drain));
+
+	/* since struct sof_ipc_stream drain only contains the reply
+	 * header there's no need for additional checks.
+	 *
+	 * we're assuming that since we're receiving the drain-related
+	 * reply then the drain operation finished successfully.
+	 */
+	if (spcm->pcm.compress)
+		snd_compr_drain_notify(stream->cstream);
```
Comment (plbossart) [inline sound/soc/sof/ipc3.c:927]: And what happens when the drain is finished? Don't we need a TRIGGER_STOP or something?

Comment (paulstelian97) [reply sound/soc/sof/ipc3.c:927]: Drain can be possible without stop if there are somehow two consecutive streams being fed, you do a partial drain for the first stream before starting the next one (useful for gapless playback I guess? Not 100% sure). I'd guess stop should wait until a drain finishes though.

File: sound/soc/sof/compress.c (hunk 1 above)
Comment (LaurentiuM1234) [reply sound/soc/sof/compress.c:292]: Yeah, most likely so

**Comment by LaurentiuM1234** [comment]
Closing this for now because it's really bad. Will reopen when I'll have a more solid solution.

**Pull request closed without merging**